File: benchmarks/comparative/implementations/python/time_bench.py
"""
Python (Brahe) time system conversion benchmarks.
"""

import brahe

from benchmarks.comparative.implementations.python.base import (
    ensure_eop,
    time_iterations,
)
from benchmarks.comparative.results import TaskResult


def _make_epochs(datetimes: list[dict]) -> list:
    """Create Epoch objects from datetime dicts."""
    epochs = []
    for dt in datetimes:
        epc = brahe.Epoch(
            dt["year"],
            dt["month"],
            dt["day"],
            dt["hour"],
            dt["minute"],
            dt["second"],
            dt["nanosecond"],
            time_system=brahe.TimeSystem.UTC,
        )
        epochs.append(epc)
    return epochs


def epoch_creation(params: dict, iterations: int) -> TaskResult:
    """Benchmark Epoch creation from datetime and JD extraction."""
    ensure_eop()
    datetimes = params["datetimes"]

    def run():
        results = []
        for dt in datetimes:
            epc = brahe.Epoch(
                dt["year"],
                dt["month"],
                dt["day"],
                dt["hour"],
                dt["minute"],
                dt["second"],
                dt["nanosecond"],
                time_system=brahe.TimeSystem.UTC,
            )
            results.append(epc.jd())
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="time.epoch_creation",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def utc_to_tai(params: dict, iterations: int) -> TaskResult:
    """Benchmark UTC to TAI conversion."""
    ensure_eop()
    datetimes = params["datetimes"]
    epochs = _make_epochs(datetimes)

    def run():
        results = []
        for epc in epochs:
            jd_tai = epc.jd_as_time_system(brahe.TimeSystem.TAI)
            results.append(jd_tai)
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="time.utc_to_tai",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def utc_to_tt(params: dict, iterations: int) -> TaskResult:
    """Benchmark UTC to TT conversion."""
    ensure_eop()
    datetimes = params["datetimes"]
    epochs = _make_epochs(datetimes)

    def run():
        results = []
        for epc in epochs:
            jd_tt = epc.jd_as_time_system(brahe.TimeSystem.TT)
            results.append(jd_tt)
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="time.utc_to_tt",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def utc_to_gps(params: dict, iterations: int) -> TaskResult:
    """Benchmark UTC to GPS conversion."""
    ensure_eop()
    datetimes = params["datetimes"]
    epochs = _make_epochs(datetimes)

    def run():
        results = []
        for epc in epochs:
            jd_gps = epc.jd_as_time_system(brahe.TimeSystem.GPS)
            results.append(jd_gps)
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="time.utc_to_gps",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def utc_to_ut1(params: dict, iterations: int) -> TaskResult:
    """Benchmark UTC to UT1 conversion."""
    ensure_eop()
    datetimes = params["datetimes"]
    epochs = _make_epochs(datetimes)

    def run():
        results = []
        for epc in epochs:
            jd_ut1 = epc.jd_as_time_system(brahe.TimeSystem.UT1)
            results.append(jd_ut1)
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="time.utc_to_ut1",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )

File: benchmarks/comparative/implementations/rust/src/time_bench.rs
use brahe::time::{Epoch, TimeSystem};
use std::time::Instant;

#[derive(serde::Deserialize)]
struct DateTimeParams {
    year: u32,
    month: u8,
    day: u8,
    hour: u8,
    minute: u8,
    second: f64,
    nanosecond: f64,
}

fn parse_datetimes(params: &serde_json::Value) -> Vec<DateTimeParams> {
    serde_json::from_value(params["datetimes"].clone()).unwrap()
}

fn make_epochs(dts: &[DateTimeParams]) -> Vec<Epoch> {
    dts.iter()
        .map(|dt| {
            Epoch::from_datetime(
                dt.year,
                dt.month,
                dt.day,
                dt.hour,
                dt.minute,
                dt.second,
                dt.nanosecond,
                TimeSystem::UTC,
            )
        })
        .collect()
}

pub fn epoch_creation(
    params: &serde_json::Value,
    iterations: usize,
) -> (Vec<f64>, serde_json::Value) {
    let dts = parse_datetimes(params);

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(dts.len());

        for dt in &dts {
            let epc = Epoch::from_datetime(
                dt.year,
                dt.month,
                dt.day,
                dt.hour,
                dt.minute,
                dt.second,
                dt.nanosecond,
                TimeSystem::UTC,
            );
            results.push(epc.jd());
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn utc_to_tai(params: &serde_json::Value, iterations: usize) -> (Vec<f64>, serde_json::Value) {
    let dts = parse_datetimes(params);
    let epochs = make_epochs(&dts);

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(epochs.len());

        for epc in &epochs {
            results.push(epc.jd_as_time_system(TimeSystem::TAI));
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn utc_to_tt(params: &serde_json::Value, iterations: usize) -> (Vec<f64>, serde_json::Value) {
    let dts = parse_datetimes(params);
    let epochs = make_epochs(&dts);

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(epochs.len());

        for epc in &epochs {
            results.push(epc.jd_as_time_system(TimeSystem::TT));
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn utc_to_gps(params: &serde_json::Value, iterations: usize) -> (Vec<f64>, serde_json::Value) {
    let dts = parse_datetimes(params);
    let epochs = make_epochs(&dts);

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(epochs.len());

        for epc in &epochs {
            results.push(epc.jd_as_time_system(TimeSystem::GPS));
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn utc_to_ut1(params: &serde_json::Value, iterations: usize) -> (Vec<f64>, serde_json::Value) {
    let dts = parse_datetimes(params);
    let epochs = make_epochs(&dts);

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(epochs.len());

        for epc in &epochs {
            results.push(epc.jd_as_time_system(TimeSystem::UT1));
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}
